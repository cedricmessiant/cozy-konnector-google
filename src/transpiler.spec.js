const transpile = require('./transpiler')

describe('Transpile from google-contacts to io.cozy-contacts', () => {
  test('Simple test', () => {
    const source = {
      resourceName: 'people/c3745049716692310153',
      etag:
        '%Eh0BAgMEBQYHCAkKCwwNDg8QERITFBUXNRk0NyIlLhoMAQIDBAUGBwgJCgsMIgxiamRQSjM1Szlmbz0=',
      metadata: {
        sources: [
          {
            type: 'CONTACT',
            id: '33f916e00d9f5889',
            etag: '#bjdPJ35K9fo=',
            updateTime: '2018-03-20T10:18:37.332001Z'
          }
        ],
        objectType: 'PERSON'
      },
      names: [
        {
          metadata: {
            primary: true,
            source: {
              type: 'CONTACT',
              id: '33f916e00d9f5889'
            }
          },
          displayName: 'Michelle Seabury',
          familyName: 'Seabury',
          givenName: 'Michelle',
          displayNameLastFirst: 'Seabury, Michelle'
        }
      ],
      photos: [
        {
          metadata: {
            primary: true,
            source: {
              type: 'CONTACT',
              id: '33f916e00d9f5889'
            }
          },
          url:
            'https://lh3.googleusercontent.com/-XdUIqdMkCWA/AAAAAAAAAAI/AAAAAAAAAAA/4252rscbv5MrelSgWD1bjXDv9XX6A51FQIhkKAU0Q____________ARinr4L4______8B/s100/photo.jpg',
          default: true
        }
      ],
      birthdays: [
        {
          metadata: {
            primary: true,
            source: {
              type: 'CONTACT',
              id: '33f916e00d9f5889'
            }
          },
          date: {
            year: 1987,
            month: 10,
            day: 5
          },
          text: '1987-10-05'
        }
      ],
      addresses: [
        {
          metadata: {
            primary: true,
            source: {
              type: 'CONTACT',
              id: '33f916e00d9f5889'
            }
          },
          formattedValue:
            '46-198 E Elford St\nGreenville, SC 29601\nÉtats-Unis ',
          streetAddress: '46-198 E Elford St',
          city: 'Greenville',
          region: 'SC',
          postalCode: '29601',
          country: 'États-Unis',
          countryCode: 'US'
        }
      ],
      emailAddresses: [
        {
          metadata: {
            primary: true,
            source: {
              type: 'CONTACT',
              id: '33f916e00d9f5889'
            }
          },
          value: 'michelle.seabury@keysoft.me'
        },
        {
          metadata: {
            source: {
              type: 'CONTACT',
              id: '33f916e00d9f5889'
            }
          },
          value: 'mitch.neal@opticast.info'
        }
      ],
      phoneNumbers: [
        {
          metadata: {
            primary: true,
            source: {
              type: 'CONTACT',
              id: '33f916e00d9f5889'
            }
          },
          value: '+33 (6)0 80 49 75 59',
          canonicalForm: '+33608049755'
        },
        {
          metadata: {
            source: {
              type: 'CONTACT',
              id: '33f916e00d9f5889'
            }
          },
          value: '+33 (6)5 59 99 73 12',
          canonicalForm: '+33655999731'
        }
      ],
      memberships: [
        {
          metadata: {
            source: {
              type: 'CONTACT',
              id: '33f916e00d9f5889'
            }
          },
          contactGroupMembership: {
            contactGroupId: 'myContacts'
          }
        }
      ]
    }

    const expected = {
      id: '0b8e8901-fa22-4b1b-bb04-2446f0a29d13',
      key: '0b8e8901-fa22-4b1b-bb04-2446f0a29d13',
      value: { rev: '1-4229b791f9a8b897937ffdece8cdee4d' },
      doc: {
        _id: '0b8e8901-fa22-4b1b-bb04-2446f0a29d13',
        _rev: '1-4229b791f9a8b897937ffdece8cdee4d',
        address: [
          {
            city: 'Greenville',
            country: 'États-Unis',
            postcode: '29601',
            primary: true,
            street: '46-198 E Elford St'
          }
        ],
        birthday: '1987-10-05',
        email: [
          { address: 'michelle.seabury@keysoft.me', primary: true },
          { address: 'mitch.neal@opticast.info', primary: false }
        ],
        name: { familyName: 'Seabury', givenName: 'Michelle' },
        phone: [
          { number: '+33 (6)0 80 49 75 59', primary: true },
          { number: '+33 (6)5 59 99 73 12', primary: false }
        ]
      }
    }

    const transpiled = {
      ...expected,
      doc: {
        ...expected.doc,
        ...transpile.toCozy(source)
      }
    }

    // first we test some internal properties
    expect(transpiled.doc.name).toEqual(expected.doc.name)
    expect(transpiled.doc.phone).toEqual(expected.doc.phone)
    expect(transpiled.doc.email).toEqual(expected.doc.email)
    expect(transpiled.doc.birthday).toEqual(expected.doc.birthday)
    expect(transpiled.doc.address).toEqual(expected.doc.address)

    // finally we test the whole snapshot
    expect(transpiled).toMatchSnapshot()
  })

  test('complete tests with `sample/*.json', () => {
    const source = require('../sample/google-contacts.json')

    const transpiled = source.map(row => ({
      doc: transpile.toCozy(row)
    }))
    expect(transpiled).toMatchSnapshot()
  })
})
