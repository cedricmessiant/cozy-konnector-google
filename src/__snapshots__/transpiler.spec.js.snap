// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transpile from google-contacts to io.cozy-contacts Simple test 1`] = `
Object {
  "doc": Object {
    "_id": "0b8e8901-fa22-4b1b-bb04-2446f0a29d13",
    "_rev": "1-4229b791f9a8b897937ffdece8cdee4d",
    "address": Array [
      Object {
        "city": "Greenville",
        "country": "États-Unis",
        "postcode": "29601",
        "primary": true,
        "street": "46-198 E Elford St",
      },
    ],
    "birthday": "1987-10-05",
    "email": Array [
      Object {
        "address": "michelle.seabury@keysoft.me",
        "primary": true,
      },
      Object {
        "address": "mitch.neal@opticast.info",
        "primary": false,
      },
    ],
    "metadata": Object {
      "google": Object {
        "metadata": Object {
          "objectType": "PERSON",
          "sources": Array [
            Object {
              "etag": "#bjdPJ35K9fo=",
              "id": "33f916e00d9f5889",
              "type": "CONTACT",
              "updateTime": "2018-03-20T10:18:37.332001Z",
            },
          ],
        },
      },
    },
    "name": Object {
      "familyName": "Seabury",
      "givenName": "Michelle",
    },
    "phone": Array [
      Object {
        "number": "+33 (6)0 80 49 75 59",
        "primary": true,
      },
      Object {
        "number": "+33 (6)5 59 99 73 12",
        "primary": false,
      },
    ],
    "resourceName": "people/c3745049716692310153",
  },
  "id": "0b8e8901-fa22-4b1b-bb04-2446f0a29d13",
  "key": "0b8e8901-fa22-4b1b-bb04-2446f0a29d13",
  "value": Object {
    "rev": "1-4229b791f9a8b897937ffdece8cdee4d",
  },
}
`;

exports[`Transpile from google-contacts to io.cozy-contacts complete tests with \`sample/*.json 1`] = `
Array [
  Object {
    "doc": Object {
      "address": Array [
        Object {
          "city": undefined,
          "country": undefined,
          "postcode": undefined,
          "primary": true,
          "street": "House Stark",
        },
      ],
      "birthday": "01/01/298",
      "email": Array [
        Object {
          "address": "jon.snow@nightswatch.com",
          "primary": true,
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#VX0fx6Lp0cw=",
                "id": "59575f3f0a9cc5c8",
                "type": "CONTACT",
                "updateTime": "2018-03-19T17:53:48.327Z",
              },
            ],
          },
        },
      },
      "name": Object {
        "familyName": "Snow",
        "givenName": "Jon",
      },
      "phone": Array [],
      "resourceName": "people/c6437718916715038152",
    },
  },
  Object {
    "doc": Object {
      "address": Array [
        Object {
          "city": undefined,
          "country": undefined,
          "postcode": undefined,
          "primary": true,
          "street": "Winterfell",
        },
        Object {
          "city": undefined,
          "country": undefined,
          "postcode": undefined,
          "primary": false,
          "street": "Winterfell",
        },
      ],
      "birthday": "01/01/263",
      "email": Array [
        Object {
          "address": "eddard.stark@stark.com",
          "primary": true,
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#PQlOlGEHd/4=",
                "id": "257de03988942c82",
                "type": "CONTACT",
                "updateTime": "2018-03-19T17:56:55.359002Z",
              },
            ],
          },
        },
      },
      "name": Object {
        "familyName": "Stark",
        "givenName": "Eddard",
      },
      "phone": Array [
        Object {
          "number": "12 34 56 78 90",
          "primary": true,
        },
      ],
      "resourceName": "people/c2701561889201335426",
    },
  },
  Object {
    "doc": Object {
      "address": Array [
        Object {
          "city": undefined,
          "country": undefined,
          "postcode": undefined,
          "primary": true,
          "street": "94 Crawford Street, 35599 Fairfield Kenya",
        },
      ],
      "birthday": undefined,
      "email": Array [
        Object {
          "address": "russell.bourquin@protheon.org",
          "primary": true,
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#n26kPNzsKTI=",
                "id": "23ddfe4b8c9b2ec1",
                "type": "CONTACT",
                "updateTime": "2018-03-20T10:16:52.748Z",
              },
            ],
          },
        },
      },
      "name": Object {
        "familyName": "Protheon",
        "givenName": "Protheon",
      },
      "phone": Array [
        Object {
          "number": "+33  (8)2 06 92 96 57",
          "primary": true,
        },
      ],
      "resourceName": "people/c2584501361615515329",
    },
  },
  Object {
    "doc": Object {
      "address": Array [
        Object {
          "city": "Greenville",
          "country": "États-Unis",
          "postcode": "29601",
          "primary": true,
          "street": "46-198 E Elford St",
        },
      ],
      "birthday": "1987-10-05",
      "email": Array [
        Object {
          "address": "michelle.seabury@keysoft.me",
          "primary": true,
        },
        Object {
          "address": "mitch.neal@opticast.info",
          "primary": false,
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#bjdPJ35K9fo=",
                "id": "33f916e00d9f5889",
                "type": "CONTACT",
                "updateTime": "2018-03-20T10:18:37.332001Z",
              },
            ],
          },
        },
      },
      "name": Object {
        "familyName": "Seabury",
        "givenName": "Michelle",
      },
      "phone": Array [
        Object {
          "number": "+33 (6)0 80 49 75 59",
          "primary": true,
        },
        Object {
          "number": "+33 (6)5 59 99 73 12",
          "primary": false,
        },
      ],
      "resourceName": "people/c3745049716692310153",
    },
  },
]
`;
